[Eric Leow Yu Quan] iP

shittake opened the pull request:
### Duke is a bot that can help you manage your to-do list and keep them in one place!

- Often find yourself **overwhelmed** and forgetting to do tasks?
- Want to have a place to _keep track of what's there to be done_?
- Afraid that you may miss out on something ~trivial~ important like your other half's birthday? 😍 

You're not alone! Scientists show that memory is much more transient than you may think 

>So, cognitive psychologists divide memory into the first 15-30 seconds, and they call this short-term memory, and alllllll the rest of memory that lasts beyond 30 seconds is long-term memory.  - ([source](https://arc.duke.edu/how-long-short-term-memory-shorter-you-might-think#:~:text=So%2C%20cognitive%20psychologists%20divide%20memory,seconds%20is%20long%2Dterm%20memory.)) 

**Features:**
- [x] Keep track of your tasks 
- [x] Allow marking tasks as complete
- [ ] Managing deadlines (Coming soon)

**Strengths**
1. Easy to use
2. Free!
3. More features to come

If you are a Java programmer, you can use it to practise Java too. Here's the ``main`` method:
``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

shittake committed:
add new message for testing

shittake committed:
Revert "add new message for testing"

This reverts commit 2b287c371910bde5fb69ce2620cb09a7831494bb.

shittake committed:
Add welcome message with introMessage method

shittake committed:
Add functionality to echo user input until user decides to end program

shittake committed:
Abstract out endMessage() method for better readability

shittake committed:
Add additional checker class and checkEnd method for user input validation

shittake committed:
Add checkListRequest method to Checker class that will check if user wishes to see the current list

shittake committed:
Add functionality to allow storage of all userinputs and print them out in order when requested by the user

shittake committed:
Add ability to print out custom message if no items have been added yet by the user

shittake committed:
Create Task class which encapsulates the name and status of each task added by the user

shittake committed:
Override toString() method in Task class for easy printing

shittake committed:
Add Task instances into the lstOfItems to capture state of each instance properly

shittake committed:
Add functionality to mark Task as completed and fix minor bugs

shittake committed:
Include message that is printed out to let user know that a task has been marked as completed

shittake committed:
Add message that tells user the number of tasks present currently after adding a valid task

shittake committed:
Create Deadline, Event and Todo classes that inherit from Task class

shittake committed:
Add functionality to add different types of Tasks into the list with their separate descriptions and timings

shittake committed:
Add Test UI Testing

shittake committed:
Create new DukeException class to handle custom exceptions

shittake committed:
Add functionality to check for empty todo command and invalid commands

shittake committed:
Add exceptions to check for incorrect formatting of Deadline and Event Tasks

shittake committed:
Add functionality to remove previously added tasks

shittake committed:
Create preliminary txt file that will later on store all information about previously added information

shittake committed:
Create new text file if it is not present in user's directory

shittake committed:
Add functionality to parse previously saved file to retrieve information

shittake committed:
Add functionality to save contents into hard disk when the user exits the session

shittake committed:
Add functionality to store deadline dates by parsing yyyy-mm-dd format and printing out in MMM dd yyyy format using the LocalDate class

shittake committed:
Merge branch 'branch-Level-7'

shittake committed:
Merge branch 'branch-Level-8'

shittake committed:
Fix bug when transforming the desired datetime format into the hard drive

shittake committed:
Fix bug where programme crashes if user did not key in any new Tasks in the first iteration of Duke

shittake committed:
Merge branch 'add-gradle-support' of https://github.com/shittake/ip

shittake committed:
Remove unnecessary imports and modify boolean variable naming to fit Java naming conventions

shittake committed:
Add additional abstraction layers

shittake committed:
Remove unnecessary lines due to abstraction

shittake committed:
Abstract out more methods to Storage and TaskList classes

shittake committed:
Add methods to Ui and Parser classes

shittake committed:
Create new duke package for all classes

shittake committed:
Add JUnit Tests for Task and TaskList classes

shittake committed:
Clean up code to fit coding standards

shittake committed:
Add JavaDocs documentation to all public methods

shittake committed:
Add functionality to search tasks given a certain keyword

shittake committed:
Merge branch 'branch-Level-9'

shittake committed:
Merge branch 'branch-A-JavaDoc'

shittake committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java

shittake committed:
Clean up merge conflicts

shittake committed:
Add basic JavaFX syntax

shittake committed:
Add functionality to echo the user

shittake committed:
Add images as avatar icons

shittake committed:
Tweak functionality to flip image of Duke when echoing commands

shittake committed:
Add ability to print out introduction message by the GUI

shittake committed:
Modify functionality to a preliminary one to echo user input only

shittake committed:
Add different styling for chat bubble from perspective of user and Duke

shittake committed:
Add functionality for most tasks in GUI

shittake committed:
Fix bug such that error messages are now displayed properly in the GUI

shittake committed:
Merge branch 'branch-Level-10'

shittake committed:
Remove unnecessary imports

markusyeo commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+
+    /**
+     * Returns the string representation of a Deadline instance in MMM d yyyy format.
+     * @return the desired string representation of a Deadline instance.
```
Perhaps you should have an empty line between description and parameter section?

shittake replied:
Thanks for pointing that out! I wasn't aware of this coding standard

markusyeo commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    /**
+     * Returns the string representation of a Deadline instance in YYYY-MM-DD format.
```
You might have missed a spacing between the descriptions and parameter section here as well.

markusyeo reviewed (commented):
LGTM! Except the formatting of your javadocs might need to follow the coding standard

mandykqh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+     */
+    public String printContents() {
+        String response = "Attempting to print out your tasks...\n";
+        if (lstOfItems.size() == 0 ) {
```
You might have added an extra space after "0".

mandykqh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    /**
+     * A method stub to test the TaskList class.
```
Perhaps you may consider starting the method header with an active verb.
Eg. "Tests the TaskList class."

KS2499 commented on src/data/duke.txt:
```diff
@@ -0,0 +1,4 @@
+3
+[T][X] yyy
+[D][X] kdayyy  (by: 2021-12-21)
+[T][X] skdakda
```
Perhaps omit this file from commit, or keep it empty so when others run duke they do not inherit your tasks?

KS2499 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,30 @@
+
+    protected String from;
+    protected String to;
+
```
Perhaps reformat `from`  and `to` to `LocalDate ` format like in deadline?

KS2499 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+        } else {
+            return false;
+        }
+    }
```
I like how you validate the task to be marked is in the correct format(integer)

mandykqh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,245 @@
+import javafx.stage.Stage;
+
+
+public class Duke extends Application {
```
Perhaps you may consider adding JavaDoc comments for your classes too!

KS2499 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+            } catch (IOException err) {
+                System.out.println(err);
+            }
+
```
Perhaps you could combine the 3 try-catch blocks into a single try-catch-catch block?

KS2499 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,30 @@
+
+public class Event extends Task {
+
+    protected String from;
+    protected String to;
```
Perhaps reformat `from` and `to` to `LocalDate` format like in Deadline class?

KS2499 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+                    isNumber = false;
+                }
+            }
+            return isNumber;
```
I like how you validate the input is in the correct format(integer).

mandykqh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+        if (currentTask.charAt(1) == 'T') {
+            task = new Todo(currentTask.substring(7));
+        } else if (currentTask.charAt(1) == 'D') {
+            String[] split = currentTask.split("by: ");
```
You may consider renaming `split` into a plural form since it is a String array.

KS2499 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+        // https://stackoverflow.com/questions/6142901/how-to-create-a-file-in-a-directory-in-java
+        if (!isExistFile()) {
+            Path desiredPath = Paths.get(filePath);
+            try {
+                Files.createDirectories(desiredPath.getParent());
+            } catch (IOException err) {
+                System.out.println(err);
+            }
+
+            try {
+                Files.createFile(desiredPath);
+            } catch (FileAlreadyExistsException err) {
+                System.err.println("already exists: " + err.getMessage());
+            } catch (IOException err) {
+                System.out.println(err);
+            }
+
+            try {
+                Files.write(Paths.get(filePath), "0".getBytes());
+            } catch (IOException err) {
+                System.out.println(err);
+            }
```
Perhaps you could combine the 3 try-catch blocks into a single try-catch-catch block?

mandykqh reviewed (commented):
Overall, your code is pretty clean aside from those few nits. Nice job! 👍

KS2499 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,58 @@
+package duke;
+
+public class Task {
```
Perhaps an alternative way this class is defined could be as an abstract class since `Tasks` are always `Todo`, `Deadline` or `Event`?

KS2499 reviewed (commented):
Overall LGTM, I really like the way you commented your javadocs and gave references to your sources when you borrowed a code idea.

KS2499 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,50 @@
+        return new Label(introString);
+    }
+
+    public String instruct() {
+        String instructString = "There are various commands you can instruct me to do.\n";
+        instructString += ("A. If you want to see your list of tasks, type 'list'\n");
+        instructString += ("B. If you want to add a task, there are three possible commands:\n");
+        instructString += ("\t1. Type 'todo (name of task)' to add a todo task.\n");
+        instructString += ("\t2. Type 'deadline (name of task) /by (date in YYYY-MM-DD)\n");
+        instructString += (" to add a task that is to be done by a certain date.\n");
+        instructString += ("\t3. Type 'event (name of task) /from (date and time) /to ");
+        instructString += ("(date and time) to add in a time-limited event.\n");
+        instructString += ("C. To delete a task, type 'delete X' where X is the task number\n");
+        instructString += ("D. To mark a task as completed, type 'mark X'\n");
+        instructString += ("E. To save and exit, type 'bye'\n");
+        return instructString;
+    }
```
I like the way you formatted your paragraph line by line. Perhaps you could look into using StringBuilder here.

KS2499 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+
+public class TaskList {
+
+    private ArrayList<Task> lstOfItems;
```
```suggestion
    private ArrayList<Task> lstOfTasks;
```

KS2499 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,58 @@
+     * Marks the task as completed.
+     * It does not matter whether the task has previously been marked as completed.
+     */
+    public void makeCompleted() {
```
Perhaps a more intuitive name like `markAsCompleted` can be used here?

KS2499 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+     * Adds and saves the user's tasks into the File at the filePath.
+     * @param tasks the user's list of tasks before termination of the chatbot.
+     */
+    public void addToFile(TaskList tasks) {
```
Any reason why you named it as such instead of `save` (like load and save)?

KS2499 reviewed (commented):
Overall LGTM, I really like the way you commented your code and gave credit to your sources!

shittake committed:
Add JavaDocs for Constructors in all Classes

Lack of documentation for constructors that may make it difficult to
understand certain parameters if not explained

Explaining the parameters used in all constructors allows for easier
appreciation of the purpose of each class, and increases ease in bug-
fixing in the future as it is easier to pinpoint which class or
attribute is responsible for the bug

shittake committed:
Clean up code with the aid of comments given by peers

Some coding standard violations such as not leaving a line in JavaDocs
between the description and the @params and @return statements

Made changes accordingly to fit the guidelines for standardization,
to ensure that JavaDocs written adhere to these standards and are of
high quality, and to maximize learning. Names of some variables were
also changed to better represent what items they were actually storing
to avoid any misconceptions.

shittake committed:
Merge pull request #1 from shittake/branch-A-CodeQuality

Clean up code with the aid of comments given by peers and improve code to adhere to code quality guidelines

shittake committed:
Merge branch 'master' of https://github.com/shittake/ip

shittake committed:
Add assertions to key parts of code

No assertions present that could lead to instances where a code does
not work properly as intended, such as attempting to access a list
that was not even created

Add assertions that kept track of assumptions made in code, to check if
these assumptions are held when certain parts of the code is run, to
avoid undefined behaviour or errors

shittake committed:
Merge branch 'master' into branch-A-Assertions

shittake committed:
Merge pull request #2 from shittake/branch-A-Assertions

Add assertions to key parts of code

shittake committed:
Merge branch 'master' of https://github.com/shittake/ip

shittake committed:
Tackle bug involving incorrect variable name due to merge conflicts

shittake committed:
Add functionality to disable further userInput upon termination

shittake committed:
Abstract method to check for termination for cleaner code

shittake committed:
Add priority level for tasks using enums

shittake committed:
Add functionality to increase/decrease priority levels for tasks, and
to allow saving and loading of previous tasks involving their respective
priority levels.

There was no ability to save previous priority levels,
causing all tasks to have low priority level every time the application
is launched

Add ability to toggle priority level incrementally (among the 3 possible
priority levels). Allow saving of data to save user time and to ensure
that users do not need to set priority level of tasks manually each
time.

shittake committed:
Add functionality to sort tasks based on priority level

shittake committed:
Merge branch 'branch-BCD-Extension'

shittake committed:
Fix bug where an extra space is added each time the list is loaded

shittake committed:
Add instructions to introductory message

shittake committed:
Fix bug where mainClassName is not "Launcher"

shittake committed:
Abstract methods for cleaner code

Long methods such as run() in Duke class and manual checking of user's
input for certain words resulted in arrow-head code and overly long
methods.

Manual checking of input was abstracted into more methods in the Parser
class, with a checking of valid integers. Code is cleaner and easier to
follow.

shittake committed:
Add more comprehensive tests for TaskListTest to account for other
methods implemented, such as unmark task and find keyword

shittake committed:
Add functionality to display tutorial again if needed

shittake committed:
Add JavaDocs for all classes

shittake committed:
Add more comprehensive test cases for TaskTest

shittake committed:
Add JavaDocs for methods in Duke class

shittake committed:
Add representative screenshot of product

shittake committed:
Add User Guide to product
